File: AOC_2024/Python/Day01/day01.py
from sys import argv
from collections import defaultdict

from Utils.utils import timer

with open(argv[1] if len(argv) > 1 else "data.txt") as file:
    data = [line.split() for line in file.read().split("\n")]
    data = (sorted([int(row[0]) for row in data]), sorted([int(row[1]) for row in data]))


def part1():
    sum = 0
    for idx, i in enumerate(data[0]):
        sum += abs(i - data[1][idx])
    return sum


def part2():
    ...
    sum = 0
    frequencies = defaultdict(int)
    for i in data[0]:
        frequencies[i] += 1
    for j in data[1]:
        if j in frequencies:
            sum += j * frequencies[j]
    return sum


if __name__ == '__main__':
    timer(part1, part2)

File: AOC_2024/C++/Day01/day01.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
#include <sstream>
#include <unordered_map>

#include "utils.h"

using namespace std;

vector<vector<int> > input_data = {{}, {}};

long long part1()
{
    long long sum = 0;
    for (int i = 0; i < input_data[0].size(); i++)
    {
        sum += abs(input_data[0][i] - input_data[1][i]);
    }
    return sum;
}

long long part2()
{
    long long sum = 0;
    unordered_map<int, int> frequencies;
    for (int i: input_data[0])
    {
        frequencies[i]++;
    }
    for (int i: input_data[1])
    {
        sum += i * frequencies[i];
    }
    return sum;
}

int main()
{
    string line;
    ifstream f("../AOC_2024/C++/Day01/data.txt");

    while (getline(f, line))
    {
        stringstream ss(line);
        int left, right;
        ss >> left >> right;
        input_data[0].push_back(left);
        input_data[1].push_back(right);
    }
    f.close();
    sort(input_data[0].begin(), input_data[0].end());
    sort(input_data[1].begin(), input_data[1].end());

    utils::timer(part1, part2);
}
